src/lib/mdns/minimal/core/FlatAllocatedQName.h

```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "QName.h"
+
+namespace mdns {
+namespace Minimal {
+
+/// A void* implementation that stores QNames as
+///   [ptr array] [name1] #0 [name2] #0 .... [namen] #0
+///
+/// Usage:
+///   - RequiredStorageSize("some", "name", "here") provides the memory block
+///   - Build(ptr, "some", "name", "here") stores all data in [ptr]
+///
+namespace FlatAllocatedQName {
+
+inline size_t RequiredStorageSize()
+{
+    return 0;
+}
+
+/// Deternmine the memory size required to store the given qname parts
+///
+/// Example:
+///   malloc(RequiredStorageSize("myhostname", "local"));
+template <typename... Args>
+inline size_t RequiredStorageSize(QNamePart name, Args &&... rest)
```

Comment:
What's the code size impact of this? Would C style varargs be smaller?

Reply:
I think it may be inlinable and the function is super small.

Godbolt shows some code: https://godbolt.org/z/5rEeMo but not overly bad (i.e. a bunch of strlen calls is for the most case).

I would prefer to optimize later if applicable.

Reply:
I have converted this to varargs and the saving is quite small on this piece of code:

```
❯ bloaty examples/all-clusters-app/esp32/build/chip-all-clusters-app.elf -- ~/tmp/chip-all-clusters-app.elf.memory_alloc_base
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +0.1% +7.65Ki  [ = ]       0    .debug_info
  +0.1% +2.07Ki  [ = ]       0    .debug_line
  +0.2% +1.65Ki  [ = ]       0    .debug_abbrev
  +0.0%     +64  [ = ]       0    .debug_frame
  +0.1%     +40  [ = ]       0    .debug_aranges
  +0.0%     +24  [ = ]       0    .debug_ranges
  +0.0%      +2  [ = ]       0    [Unmapped]
  -0.0%     -39  [ = ]       0    .strtab
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName19RequiredStorageSizeIJRA6_KcEEEjPS3_DpOT_
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName8Internal10InitializeIJRA5_KcRA6_S4_EEEvPPS4_PcS9_DpOT_
  [DEL]     -76  [ = ]       0    .xt.prop._ZN4mdns7Minimal18FlatAllocatedQName8Internal10InitializeIJRA6_KcEEEvPPS4_PcS7_DpOT_
  -0.1%     -80  [ = ]       0    .symtab
 -16.5%    -120  [ = ]       0    .xt.lit._ZN4mdns7Minimal14ResponseSenderD5Ev
  -0.0%    -124  -0.0%    -124    .flash.text
  -1.5%    -563  [ = ]       0    .shstrtab
  -0.1%   -1017  [ = ]       0    .debug_loc
  -0.1% -1.72Ki  [ = ]       0    .debug_str
  +0.0% +7.66Ki  -0.0%    -124    TOTAL
```

however I expect it may grow as we use this method more (unclear how much though, so I expect we will often use it with just 2/3/4 arguments).

Since varargs loses typesafety I think we should defer this optimization for later.